internal/cloud/azure/application_permission.go

```diff
@@ -20,132 +23,117 @@ type ApplicationPermissionConfig struct {
+
+	raClient.Authorizer = authorizer
 	return raClient, nil
 }
 
 type RoleDefinitionsClient interface {
-	Get(ctx context.Context, scope string, roleDefinitionID string, options *armauthorization.RoleDefinitionsClientGetOptions) (armauthorization.RoleDefinitionsClientGetResponse, error)
-	NewListPager(scope string, options *armauthorization.RoleDefinitionsClientListOptions) *runtime.Pager[armauthorization.RoleDefinitionsClientListResponse]
-	//List(ctx context.Context, query odata.Query) (*[]msgraph.UnifiedRoleDefinition, int, error)
+	GetByID(ctx context.Context, roleID string) (result authorization.RoleDefinition, err error)
+	List(ctx context.Context, scope string, filter string) (result authorization.RoleDefinitionListResultPage, err error)
 }
 
 func (c *AzureConfig) NewRoleDefinitionsClient(ctx context.Context) (RoleDefinitionsClient, error) {
-	creds, credsErr := azidentity.NewClientSecretCredential(
-		c.Provider.TenantID.Value,
-		c.Provider.ClientID.Value,
-		c.Provider.ClientSecret.Value,
-		&azidentity.ClientSecretCredentialOptions{})
-	if credsErr != nil {
-		return nil, credsErr
-	}
-	//rdClient := authorization.NewRoleDefinitionsClient()//NewRoleDefinitionsClient(c.Provider.SubscriptionID.Value)
-	rdClient, clientErr := armauthorization.NewRoleDefinitionsClient(creds, &arm.ClientOptions{})
-	if clientErr != nil {
-		return nil, clientErr
-	}
-	// authorizer, authorizerErr := c.authConfig.NewAuthorizer(ctx, c.authConfig.Environment.ResourceManager)
-	// if authorizerErr != nil {
-	// 	return nil, authorizerErr
-	// }
-	//rdClient := msgraph.NewRoleDefinitionsClient(c.Provider.TenantID.Value)
+	rdClient := authorization.NewRoleDefinitionsClient(c.Provider.SubscriptionID.Value)
+
+	authorizer, err := getAzureResourceManagerAuthorizer(ctx, c)
+	if err != nil {
+		return nil, fmt.Errorf("error creating RoleDefinitionsClient: %+v", err)
+	}
+
+	rdClient.Authorizer = authorizer
 	return rdClient, nil
 }
 
 func CreateApplicationPermission(ctx context.Context, config *ApplicationPermissionConfig, raClient RoleAssignmentsClient, rdClient RoleDefinitionsClient) error {
 
-	filter := fmt.Sprintf("displayName eq '%s'", config.RoleName)
-	rdPager := rdClient.NewListPager(config.Scope,
-		&armauthorization.RoleDefinitionsClientListOptions{
-			Filter: &filter})
-	//roleDefinitions, _, err := rdClient.List(ctx, odata.Query{ //})
-	//Filter: fmt.Sprintf("displayName eq '%s'", config.RoleName),
-	//Filter: fmt.Sprintf("roleName eq '%s'", config.RoleName),
-	//})
-	// if err != nil {
-	// 	return fmt.Errorf("loading Role Definition List: %+v", err)
-	// }
-	for rdPager.More() {
-		nextResult, nextErr := rdPager.NextPage(ctx)
-		if nextErr != nil {
-			return nextErr
-		}
-		for _, v := range nextResult.Value {
-			log.Default().Printf("\nDisplayName: %v\t\t\tID: %v\n\n", *v.Properties.RoleName, *v.ID)
-		}
+	roleDefinitions, err := rdClient.List(ctx, config.Scope, fmt.Sprintf("roleName eq '%s'", config.RoleName))
```

Comment:
So this is a strange Azure-ism...Roles don't seem to be "singleton".  They vary based off of the "scope".  So we have to do a lookup to get the Role ID by looking up the RoleName (ex `Storage Account Contributor`) with a scope.

Technically we *could* change up our security block and have the bundle output the **fully scoped** ID of the role (`/subscriptions/sdkfjldf/Microsoft.something/lskdjfskldj/Roles/sldkfsldkjf/Whatever`) and we might be able to skip this step, but I don't know if that would play nicely with other clients.  I pulled all this code from their own terraform provider, so it seems they kinda expect to jump through these hoops on Role lookup.

Reply:
great context

Reply:
I think it's a good idea to mimic their provider and leave the security blocks as-is for now.